psh/date/date.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Phoenix-RTOS
+ *
+ * date - print or set the system date and time
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Mateusz Niewiadomski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <sys/time.h>
+#include <unistd.h>
+
+#include "../psh.h"
+
+
+enum mode { SKIP,
+	HELP,
+	PRINT,
+	WRITE };
+
+static void psh_date_info(void)
+{
+	printf("print/set the system date and time");
+}
+
+
+static void psh_date_help(const char *prog)
+{
+	printf("Usage: %s [-h] [-s EPOCH] [+FORMAT]\n", prog);
+	printf("  -h:  shows this help message\n");
+	printf("  -s:  set system time described by EPOCH (POSIX time format)\n");
+	printf("  FORMAT: string with POSIX date formatting characters\n");
+	printf("NOTE: FORMAT string not supported with '-s' option\n");
+}
+
+
+static int psh_date_print(char *format)
+{
+	time_t t;
+	struct tm *tinfo;
+	char defformat[] = "+%a, %d %b %y %H:%M:%S";
+	char buf[64];
+
+	if (format == NULL) {
+		format = defformat;
+	}
+	else if (format[0] != '+') {
+		fprintf(stderr, "date: invalid date '%s'\n", format);
+		return 1;
+	}
+
+	time(&t);
+	tinfo = localtime(&t);
+
+	/* passing `&format[1]` for omission of '+' at the beginning */
+	if (strftime(buf, 64, &format[1], tinfo) == 0) {
+		fprintf(stderr, "date: '%s' expansion too long\n", format);
+		return 1;
+	}
+	printf("%s\n", buf);
+
+	return 0;
+}
+
+
+static int psh_date_set(char *timestring, const char *format)
+{
+	struct timeval tv;
+	struct timezone tz = { .tz_dsttime = 0, .tz_minuteswest = 0 };
+	char *end;
+
+	/* FIXME: no strptime() implementation in libphoenix */
+	if (format != NULL) {
+		fprintf(stderr, "date: [-s] does not support FORMAT\n");
+		return 1;
+	}
+
+	if (timestring == NULL) {
+		return 1;
+	}
+
+	tv.tv_sec = 0;
+	tv.tv_sec = strtoull(timestring, &end, 10);
+	if ((end == timestring) || (*end != '\0') || ((long long int)tv.tv_sec < 0)) {
+		fprintf(stderr, "Uncrecognized argument '%s'!\n", timestring);
+		return 1;
+	}
+
+	return settimeofday(&tv, &tz);
+}
+
+
+static int psh_date(int argc, char **argv)
+{
+	int c, maxargc;
+	int status = 0;
+	char *timestring = NULL, *format = NULL;
+	enum mode m = PRINT;
+
+	/* no getopt loop, only one argument expected */
+	c = getopt(argc, argv, "hs:");
+	switch (c) {
+		case 'h':
+			m = HELP;
+			maxargc = 2;
+			break;
+
+		case 's':
+			m = WRITE;
+			timestring = optarg;
+			maxargc = 4;
+			break;
+
+		case -1:
+			m = PRINT;
+			maxargc = 2;
+			break;
+
+		default:
+			m = SKIP;
+			break;
+	}
+
+	/* evaluate additional arguments/format */
+	if (m != SKIP) {
+		if (argc > maxargc) {
```

Comment:
`maxargc` should not be needed, just checking current `optind` value against `argc`. Probably You needed it because You've removed the `while(getopt())` so `optind` was not updated correctly in case `-s` option was passed